### PR TITLE
Core: Reimplementation of the Warden anticheat.

### PR DESCRIPTION
**Pull request by Warpten**
**Changes proposed**:

- Implemented support for x86 and x86-64 Windows modules.
   x86-64 completely untested, implemented via an early Legion module.
- Modernize some of the cryptographic code in the meantime.

**Details**:

* Scan type is now defined by a custom enum.
* Each scan is specific to a platform.
* Scans store query data in `data0` and `data1` respectively, depending on the scan type itself.

**Tests performed**: (Does it build, tested in-game, etc)

- [x] PCH Build
- [x] Module check.
- [x] Memory check.
- [ ] Page check (not sure how to use this yet)
- [x] Lua check.
- [ ] Proc check (not sure how to use this yet)
- [x] File check

**Known issues and TODOs**

- [x] x86 MPQ Checks cause client crashes.
  This module thrashes the **esi** register when computing a checksum of the file.
  We can't fix it.
- [ ] Add result handling for procedure checks.

### CODE REVIEW

**Comment by limoncin** [comment]
hello, great job!

Excuse me, but I don't quite understand the code, this reinterpretation that you get?
detect hacks like hitchhiker's hack?
I also see something about lua, does it detect PQR?
Could this reinterpretation be passed to branch 3.3.5?

**Comment by Warpten** [comment]
??
no
yes
no

File: src/server/game/Warden/Checks/WardenFileCheck.cpp [hunk 1]
```diff
@@ -93,3 +96,57 @@ bool WardenFileCheck::ProcessResponse(Warden* warden, ByteBuffer& packet) const
...
+        case LOCALE_esES:
+            if (!checkFlags.HasFlag(WardenCheckFlags::SpanishLocale))
+                return false;
+            break;
+        case LOCALE_esMX:
+            if (!checkFlags.HasFlag(WardenCheckFlags::MexicanLocale))
+                return false;
+            break;
+        case LOCALE_ruRU:
+            if (!checkFlags.HasFlag(WardenCheckFlags::BlyatLocale))
```
Comment (Ovahlord) [inline src/server/game/Warden/Checks/WardenFileCheck.cpp:139]: really?

Comment (Warpten) [reply src/server/game/Warden/Checks/WardenFileCheck.cpp:139]: Let a man have his fun :(

**Comment by Warpten** [comment]
Not forgotten, working on tools related to this

**Comment by Keader** [comment]
any remote change to you take a look in https://github.com/TrinityCore/TrinityCore/issues/23035 @Warpten ?

btw, hi @Ovahlord sucker <3

**Comment by Warpten** [comment]
No

File: sql/base/auth_database.sql [hunk 2]
```diff
@@ -2008,7 +2033,7 @@ INSERT INTO `updates` VALUES
 ('2020_02_19_00_auth.sql','B3738FB3FA1C75E430EC64B5E9635A5E52E3AB36','ARCHIVED','2020-02-13 16:37:42',0),
 ('2020_05_19_00_auth.sql','ED98E556D7B1F7A62792B47C56AFB5019135D10F','ARCHIVED','2020-05-19 01:44:34',0),
 ('2020_07_22_00_auth.sql','A5D4EC8FCFAB4F2DCE70EDCAD1ACBFB484FD68D5','RELEASED','2020-07-22 17:04:47',0);
```
Comment (funjoker) [inline sql/base/auth_database.sql:2035]: make it ), at the end

Comment (Warpten) [reply sql/base/auth_database.sql:2035]: oh. kek

**Comment by Warpten** [comment]
No clue why Travis is failing.

**Comment by Warpten** [comment]
Atually i forgot to check some stuff

**Comment by Ovahlord** [comment]
Travis builds with boost 1.67. Dunno if has_value was a thing therr yet

**Comment by funjoker** [comment]
had_value was introduced in 1.68.
So I guess it’s time to drop 1.67 :P

File: sql/base/dev/world_database.sql [hunk 3]
```diff
@@ -3829,15 +3829,14 @@ CREATE TABLE `warden_checks` (
...
-  `platform` varchar(5) NOT NULL,
   `data0` varchar(255) DEFAULT NULL,
   `data1` varchar(255) DEFAULT NULL,
   `address` int(10) UNSIGNED NULL DEFAULT NULL,
   `length` tinyint(3) UNSIGNED NULL DEFAULT NULL,
   `result` varchar(24) DEFAULT NULL,
   `comment` text DEFAULT NULL,
   PRIMARY KEY (`id`)
-) ENGINE=MyISAM AUTO_INCREMENT=787 DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci;
+) ENGINE=MyISAM DEFAULT CHARSET=utf8;
```
Comment (funjoker) [inline sql/base/dev/world_database.sql:3839]: can we stick to utf8mb4 ? :)

Comment (Warpten) [reply sql/base/dev/world_database.sql:3839]: Sure, whatever, that's the rebase. Thought I'd update the branch, don't really mean to finish it yet.

Comment (funjoker) [reply sql/base/dev/world_database.sql:3839]: yeah, no rush
just want to keep the proper utf8 here

**Comment by Ovahlord** [comment]
so what's the status of that one here?

File: sql/updates/world/4.3.4/9999_99_99_01_world.sql [hunk 4]
```diff
@@ -0,0 +1,78 @@
+-- Keys (pregenerated)
+DROP TABLE IF EXISTS warden_keys;
+CREATE TABLE `warden_keys`  (
+  `id` int unsigned NOT NULL AUTO_INCREMENT,
+  `module` int unsigned NOT NULL,
+  `seed` text NOT NULL,
+  `clientKey` text NOT NULL,
+  `serverKey` text NOT NULL,
+  PRIMARY KEY (`id`)
+) ENGINE=MyISAM DEFAULT CHARSET=utf8 ROW_FORMAT=FIXED;
```
Comment (Ovahlord) [inline sql/updates/world/4.3.4/9999_99_99_01_world.sql:10]: InnoDB, utf8mb4

File: sql/updates/world/4.3.4/9999_99_99_01_world.sql [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+  `driverCheckCode` TINYINT UNSIGNED,
+  `luaStringCheckCode` TINYINT UNSIGNED,
+  `luaExecCheckCode` TINYINT UNSIGNED,
+  `filesystemCheckCode` TINYINT UNSIGNED,
+  `procCheckCode` TINYINT UNSIGNED,
+  `timingCheckCode` TINYINT UNSIGNED,
+  `interopCheckCode` TINYINT UNSIGNED,
+  `comment` VARCHAR(255),
+  PRIMARY KEY (`id`)
+) ENGINE=MyISAM DEFAULT CHARSET=utf8 ROW_FORMAT=FIXED;
```
Comment (Ovahlord) [inline sql/updates/world/4.3.4/9999_99_99_01_world.sql:74]: u know the drill

File: sql/updates/world/4.3.4/9999_99_99_00_world.sql [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+  `enabled` tinyint(1) NOT NULL DEFAULT 0,
+  `data0` varchar(255) DEFAULT NULL,
+  `data1` varchar(255) DEFAULT NULL,
+  `address` int(10) UNSIGNED NULL DEFAULT NULL,
+  `length` tinyint(3) UNSIGNED NULL DEFAULT NULL,
+  `result` varchar(40) DEFAULT NULL,
+  `comment` text DEFAULT NULL,
+  `VerifiedBuild` SMALLINT(5) NOT NULL DEFAULT 0,
+  PRIMARY KEY (`id`)
+) ENGINE=MyISAM DEFAULT CHARSET=utf8;
```
Comment (Ovahlord) [inline sql/updates/world/4.3.4/9999_99_99_00_world.sql:15]: same here

File: sql/base/dev/world_database.sql [hunk 7]
```diff
@@ -3839,16 +3839,54 @@ DROP TABLE IF EXISTS `warden_checks`;
...
+  `data0` varchar(255) DEFAULT NULL,
+  `data1` varchar(255) DEFAULT NULL,
+  `address` int(10) UNSIGNED NULL DEFAULT NULL,
+  `length` tinyint(3) UNSIGNED NULL DEFAULT NULL,
+  `result` varchar(40) DEFAULT NULL,
+  `comment` text DEFAULT NULL,
+  `VerifiedBuild` SMALLINT(5) NOT NULL DEFAULT 0,
   PRIMARY KEY (`id`)
-) ENGINE=InnoDB AUTO_INCREMENT=787 DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci;
+) ENGINE=MyISAM DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci;
```
Comment (Ovahlord) [inline sql/base/dev/world_database.sql:3854]: innodb

**Comment by Warpten** [comment]
Insert something about maybe picking this up since I got a new pc

**Comment by Ovahlord** [comment]
Insert something about when?

**Comment by Ovahlord** [comment]
soooo?

**Comment by Warpten** [comment]
Are you OK with scrapping the cataclysm module entirely and just using the 32 bits legion module for the sake of having working checks?

**Comment by funjoker** [comment]
Im fine since 32bit stuff is not available anymore.

**Comment by Ovahlord** [comment]
It's fine

**Comment by Ovahlord** [comment]
wtf is going with github... I can't re-open PRs and it nuked almost all of them

**Comment by funjoker** [comment]
Let me try :D

**Comment by funjoker** [comment]
@Warpten you may open a new PR? xD

**Comment by funjoker** [comment]
TCPP is not linked to TrinityCore as fork.
Your TC repo is a fork of TrinityCore, but not TCPP.
So i guess that's why it is closed now.
Only forks can have PRs

**Comment by Warpten** [comment]
That is fucking stupid

**Comment by funjoker** [comment]
GitHub thingy :D

**Pull request closed without merging**